adapter: add Postgres-backed TimestampOracle

aljoscha opened the pull request:
Part of #22029 
Design document: https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20230921_distributed_ts_oracle.md

This adds an initial, working version of a postgres-backed timestamp
oracle (really, it's CRDB-backed, of course...).

This adds:

- the new timestamp oracle implementation
- a system var `timestamp_oracle` for switching between the new
  'postgres' oracle and the (retroactively named) 'catalog' oracle
- a new (optional) argument to `environmentd`: `--timestamp-oracle-url`,
  that specifies the postgres connection string (including database and
  schema) where the oracle should be stored.

This is not yet meant to be used in production! Before that, we need to
address at least:

- add internal retries
- add metrics
- add optimizations that are required because oracle operations are now
  more expensive: batching of `read_ts` calls, improved batching of
  group commits (which do `write_ts`-related operations)
- involve QA, figure out any testing/nightly testing/rollout concerns
- teach cloud infra about the newly required postgres url for the
  oracle, including setting up the required database/schema

### Tips for reviewer

Commits have meaningful text, plus there's explanatory comments in the code. I left some WIP questions (largely for @danhhz because he has most context).

I left in the TODOs about retries, and some of the other things mentioned above so that it's easier to review the bare minimum structure of the new oracle code. I will open follow-up PRs for those.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

aljoscha commented:
@pH14 This adds the new, optional `--timestamp-oracle-url` argument to `environmentd`. I'll cut an issue in the cloud repo once this is merged.

philip-stoev commented on src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:
```diff
@@ -0,0 +1,386 @@
+
+    #[tracing::instrument(name = "oracle::apply_write", level = "debug", skip_all)]
+    async fn apply_write(&mut self, write_ts: Timestamp) {
+        let q = r#"
```
The query from the design document for `apply_write` is 

````
UPDATE
    timestamp_oracle SET write_ts = GREATEST(write_ts, $write_ts),
    read_ts = GREATEST(read_ts, write_ts)
    WHERE timeline = $timeline;
```

I think the query from the design document is suspect, `write_ts` that is referenced in ` read_ts = GREATEST(read_ts, write_ts)` will actually be the old `write_ts` value, the one from before the ` write_ts = GREATEST(write_ts, $write_ts)` part of the update.

philip-stoev replied:
I am not qualified unfortunately to say as I do not know enough about timestamp advancement. I just wanted to flag the discrepancy.

danhhz replied:
Calling @jkosh44, I have far less context on specifically the guarantee we need here.

It's possible that neither of them are correct if the old read_ts and the new write_ts are both less than the old write_ts? E.g. if the old read_ts is 1, the new write_ts is 2, and the old write_ts is 3, then we'd leave the write_ts at 3 and only advance the read_ts to 2. Is that what we want?

Regardless of the outcome, I think this file wants some unit tests a la the persist ones that are gated behind an env var with a postgres url

danhhz replied:
Oooh, we should copy Joe's 3 properties and 3 scenarios/outcomes in as a comment :)

philip-stoev commented:
Item No 1 . After enabling this by default for mzcompose,

```
diff --git a/misc/python/materialize/mzcompose/__init__.py b/misc/python/materialize/mzcompose/__init__.py
index 25b4f15595..04382254a1 100644
--- a/misc/python/materialize/mzcompose/__init__.py
+++ b/misc/python/materialize/mzcompose/__init__.py
@@ -69,6 +69,7 @@ DEFAULT_SYSTEM_PARAMETERS = {
     "persist_streaming_snapshot_and_fetch_enabled": "true",
     "enable_unified_clusters": "true",
     "enable_jemalloc_profiling": "true",
+    "timestamp_oracle": "postgres"
 }
```

I get the following quick crash with testdrive:

```
cd test/testdrive
./mzcompose down -v ; ./mzcompose run default 
...
avro-cdcv2.td:109:1: error: executing query failed: connection closed
```

```
environmentd: 2023-10-10T08:44:31.899396Z  INFO mz_adapter::coord::timeline: opening a new Postgres TimestampOracle for timeline External("materialize.public.data_schema_inline")
environmentd: 2023-10-10T08:44:31.899418Z  INFO mz_adapter::coord::timestamp_oracle::postgres_oracle: opening PostgresTimestampOracle url="postgresql://root@ea4b2988e387:26257?options=--search_path=timestamp_oracle"
thread 'coordinator' panicked at 'called `Result::unwrap()` on an `Err` value: AlreadyReg', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ad39b9e651b723c8-1/materialize/tests/src/ore/src/metrics.rs:229:58
```

philip-stoev commented on src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:
```diff
@@ -0,0 +1,386 @@
+        let client = self
+            .get_connection()
+            .await
+            .expect("todo: add internal retries");
```
This one too is being hit in the CI

```
platform-checks-materialized-1  | coordinator' panicked at 'todo: add internal retries: Backend(Error { kind: Connect, cause: Some(Custom { kind: Uncategorized, error: "failed to lookup address information: Name or service not known" }) })cluster-s2-replica-s2: 2023-10-10T10:29:35.251579Z  INFO mz_persist_client::internal::machine: external operation fetch_state::scan failed, retrying in 64ms: indeterminate: error connecting to server: failed to lookup address information: Name or service not known
platform-checks-materialized-1  | ', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ad39b9e651b723c8-1/materialize/tests/src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:318:14
```

In other words, I think we can rely on the existing tests once all the retry logic has been implemented.

aljoscha commented:
> Item No 1 . After enabling this by default for mzcompose,
> 
> ```
> diff --git a/misc/python/materialize/mzcompose/__init__.py b/misc/python/materialize/mzcompose/__init__.py
> index 25b4f15595..04382254a1 100644
> --- a/misc/python/materialize/mzcompose/__init__.py
> +++ b/misc/python/materialize/mzcompose/__init__.py
> @@ -69,6 +69,7 @@ DEFAULT_SYSTEM_PARAMETERS = {
>      "persist_streaming_snapshot_and_fetch_enabled": "true",
>      "enable_unified_clusters": "true",
>      "enable_jemalloc_profiling": "true",
> +    "timestamp_oracle": "postgres"
>  }
> ```
> 
> I get the following quick crash with testdrive:
> 
> ```
> cd test/testdrive
> ./mzcompose down -v ; ./mzcompose run default 
> ...
> avro-cdcv2.td:109:1: error: executing query failed: connection closed
> ```
> 
> ```
> environmentd: 2023-10-10T08:44:31.899396Z  INFO mz_adapter::coord::timeline: opening a new Postgres TimestampOracle for timeline External("materialize.public.data_schema_inline")
> environmentd: 2023-10-10T08:44:31.899418Z  INFO mz_adapter::coord::timestamp_oracle::postgres_oracle: opening PostgresTimestampOracle url="postgresql://root@ea4b2988e387:26257?options=--search_path=timestamp_oracle"
> thread 'coordinator' panicked at 'called `Result::unwrap()` on an `Err` value: AlreadyReg', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ad39b9e651b723c8-1/materialize/tests/src/ore/src/metrics.rs:229:58
> ```

yep, this is a problem with how I'm setting up metrics. I have a follow-up change which fixes that and adds more metrics. I might push that on top of this PR to get rid of the problem.

danhhz commented on src/adapter/src/coord/timestamp_oracle/metrics.rs:
```diff
@@ -0,0 +1,159 @@
+                help: "time spent in oracle operations",
+                var_labels: ["op"],
+            )),
+            external_op_latency: registry.register(metric!(
```
Looks like these are all starting as false (which I agree with), so let's pull just remove the `external_op_latency` stuff.

aljoscha replied:
done :ok_hand:

danhhz commented on src/adapter/src/coord/timestamp_oracle/metrics.rs:
```diff
@@ -0,0 +1,159 @@
+                help: "count of oracle operations succeeded",
+                var_labels: ["op"],
+            )),
+            _external_op_failed: registry.register(metric!(
```
We're expecting this to be used once we add retries?

aljoscha replied:
yep!

danhhz commented on misc/images/materialized/entrypoint.sh:
```diff
@@ -40,6 +40,7 @@ cockroach sql --insecure -e "SET CLUSTER SETTING sql.stats.forecasts.enabled = f
 cockroach sql --insecure -e "CREATE SCHEMA IF NOT EXISTS consensus"
 cockroach sql --insecure -e "CREATE SCHEMA IF NOT EXISTS storage"
 cockroach sql --insecure -e "CREATE SCHEMA IF NOT EXISTS adapter"
+cockroach sql --insecure -e "CREATE SCHEMA IF NOT EXISTS timestamp_oracle"
```
bikeshed: Very much a judgement call, but personally I'd be inclined to make this shorter (one word). Maybe something like "timestamps" or "oracle". Maaaaayyyybe "tsoracle" if we think those are too ambiguous

aljoscha replied:
you mean only the schema here or the naming of the parameter and things in other places?

danhhz replied:
Mostly the things in external systems (e.g. crdb schema)

aljoscha replied:
This one here is only used for testing, and the naming of the real one will be up to the cloud team, but I went ahead and shortened it anyways. :ok_hand:

danhhz commented on src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:
```diff
@@ -0,0 +1,332 @@
+// collect these stats fill up the jobs table (slowing down all sorts of
+// things).
+//
+// WIP: Do we need this for the postgres oracle? This is originally from
```
Assuming the "queries are single-table and tuned to hit (a specific) index" bit is true, which I believe it is, I'd keep this, yeah. The stats jobs will be much less expensive here, since we are just reusing the same row and modifying it, but they don't get us anything and.... I don't trust them now :)

danhhz replied:
looks like this WIP is still here?

danhhz commented on src/adapter/src/coord.rs:
```diff
@@ -2081,13 +2098,79 @@ pub async fn serve(
+    // We get the timestamp oracle impl once on startup, to ensure that it
+    // doesn't change in between when the system var changes: all oracles must
+    // use the same impl!
+    let timestamp_oracle_impl = catalog.system_config().timestamp_oracle_impl();
```
ooc, did you check that this works like you'd expect (we save the change when LD updates and pick it up on the next restart)? Seems easy enough to verify in a staging env with a build of this PR

danhhz replied:
ping :)

aljoscha replied:
I didn't check it with LD, no. I imagine it'll be easy enough to do, though, but I'll first try and get through the rest of the issues :see_no_evil:

danhhz replied:
okay no biggie, worst case we can fix it up in post if it doesn't work the way we think

danhhz commented on src/adapter/src/coord.rs:
```diff
@@ -2081,13 +2098,79 @@ pub async fn serve(
     let coord_now = now.clone();
     let advance_timelines_interval = tokio::time::interval(catalog.config().timestamp_interval);
+
+    // We get the timestamp oracle impl once on startup, to ensure that it
```
Let's pull this whole block out into a method :)

aljoscha replied:
good call! I've gotten too used to methods being several pages long :see_no_evil:

danhhz commented on src/adapter/src/coord/timeline.rs:
```diff
@@ -190,34 +197,67 @@ impl Coordinator {
-                    timestamp_persistence,
-                )
-                .await
+                NOW_ZERO.clone()
```
I realize this is effectively code movement, but seems like a nice time to add a comment on what's going on here

aljoscha replied:
will do, and I'll also change this to use `Timestamp::minimum`, like before, to make the intent clearer

danhhz commented on src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:
```diff
@@ -0,0 +1,399 @@
+
+        let client = postgres_client.get_connection().await?;
+
+        // WIP: Do we need this for the pg oracle? This is originally from
```
Maybe! The comment isn't quite true of this impl (this modifies a small set of rows instead of creating and deleting new rows) but I think GC would still be an issue. The main disadvantage to setting this lower is that we can't use incremental backups, but not sure that gets us a ton here anyway. (A secondary disadvantage is that it's a pretty atypical setting to use and will likely get pointed out by CRL support and have to be explained if/when we escalate anything, but we're already getting this one because persist.)

aljoscha replied:
I'll leave it in then and remove the WIP comment

danhhz commented on src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:
```diff
@@ -0,0 +1,399 @@
+            .await
+            .expect("todo: add internal retries");
+        let _ = client
+            .execute(&statement, &[&oracle.timeline, &0i64, &0i64])
```
I'd be more comfortable if we used `initially` here instead of the `0`s. Any reason not to do that?

aljoscha replied:
done :ok_hand:

danhhz commented on src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:
```diff
@@ -0,0 +1,399 @@
+        PostgresTimestampOracleConfig {
+            url: url.to_string(),
+            metrics,
+            // NOTE: These defaults are taken as is from the persist Consensus
```
For sure it doesn't have to be this PR, but we 100% need to make them dynamic here too. Ideally before turning this on in prod.

danhhz replied:
Ping! Can we make this at least a TODO or something, to try to help us not forget

aljoscha replied:
will do!

danhhz commented on src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:
```diff
@@ -0,0 +1,399 @@
+        let write_ts: i64 = result.try_get("write_ts").expect("missing column write_ts");
+        let write_ts = Timestamp::try_from(write_ts).expect("we don't use negative timestamps");
+
+        debug!(write_ts = ?write_ts, "write_ts");
```
Maybe something more descriptive like "returning write_ts", here and elsewhere. I'd also include any arguments to the query in the log message

aljoscha replied:
will do :ok_hand:

danhhz commented on src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:
```diff
@@ -0,0 +1,399 @@
+}
+
+// A wrapper around the `fallible_` methods that adds operation metrics and
+// (TODO: future work) retries.
```
Given that CI is catching panics with this on, I'm inclined to pull the retries into this PR. Should be pretty straightforward?

aljoscha replied:
I'll look into that once I'm through with the other things

danhhz commented on src/sql/src/session/vars.rs:
```diff
@@ -656,6 +656,13 @@ const PERSIST_FAST_PATH_LIMIT: ServerVar<usize> = ServerVar {
 
+const TIMESTAMP_ORACLE_IMPL: ServerVar<TimestampOracleImpl> = ServerVar {
+    name: UncasedStr::new("timestamp_oracle"),
+    value: &TimestampOracleImpl::Catalog,
```
Heh, I'd have used an `enable_postgres_timestamp_oracle` bool to make this more like the normal feature flags, but this works too :)

aljoscha replied:
it's not enough, no, and we need the commandline arg to, say, initialize the catalog oracle from the postgres oracle, in case we're switching back. And the idea was to add the commandline arg (and have cloud deal with adding the schema/database for it), and we'd have the LD flag for then enabling or disabling the oracle at will.

Also, and here I could be wrong and haven't set things up correctly: I think normally you wouldn't use `ALTER SYSTEM` but change it in LD.

danhhz reviewed (commented):
Looking good!

jkosh44 commented on src/adapter/src/coord.rs:
```diff
@@ -2081,13 +2098,79 @@ pub async fn serve(
+    // This mostly works, but there can be linearizability violations, because
+    // there is no central moment where do distributed coordination for both
+    // oracle types. Working around this seems prohibitively hard, maybe even
+    // impossible so we have to life with this window of potential violations
```
```suggestion
    // impossible so we have to live with this window of potential violations
```

jkosh44 commented on src/adapter/src/coord.rs:
```diff
@@ -2107,6 +2110,63 @@ pub async fn serve(
+            let entry = initial_timestamps
+                .entry(Timeline::from_str(&timeline).expect("could not parse timeline"));
+
+            entry
+                .and_modify(|current_ts| *current_ts = std::cmp::max(*current_ts, ts))
+                .or_insert(ts);
```
Do we need to re-persist all of these timestamps to the active oracle in case some of them changed? Or is that done later in `ensure_timeline_state_with_initial_time`?

jkosh44 commented on src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:
```diff
@@ -204,6 +204,60 @@ where
+            return Ok(Vec::new());
+        }
+
+        let q = r#"
+            SELECT timeline, GREATEST(read_ts, write_ts) as ts FROM timestamp_oracle;
+        "#;
+        let statement = client.prepare(q).await?;
+        let rows = client
+            .query(&statement, &[])
+            .await
+            .expect("todo: add internal retries");
```
Is it worth doing this in a transaction with the previous statement? I can't think of a scenario off the top of my head where someone would delete the `timestamp_oracle` table, but I don't think it will be overly expensive.

aljoscha replied:
will do!

jkosh44 reviewed (approved):
Nice!

aljoscha commented on src/adapter/src/coord.rs:
```diff
@@ -2107,6 +2110,63 @@ pub async fn serve(
+    // that initializes oracles on bootstrap, once we have fully migrated to the
+    // new postgres/crdb-backed oracle.
+    let catalog_oracle_timestamps = catalog.get_all_persisted_timestamps().await?;
+    let debug_msg = catalog_oracle_timestamps
```
done :ok_hand:

aljoscha commented on src/adapter/src/coord/timeline.rs:
```diff
@@ -190,28 +198,62 @@ impl Coordinator {
+                            now_fn,
+                        )
+                        .await
+                        .expect("todo"),
```
:see_no_evil:

aljoscha commented on src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:
```diff
@@ -0,0 +1,332 @@
+    read_ts bigint NOT NULL,
+    write_ts bigint NOT NULL,
+    PRIMARY KEY(timeline)
+) WITH (sql_stats_automatic_collection_enabled = false);
```
Yeah, same, but I chose to use "postgres" because that's what all the other places use. I think I left a snarky comment in a commit message somewhere ... :sweat_smile:

danhhz commented on misc/cockroach/setup_materialize.sql:
```diff
@@ -18,3 +18,4 @@ SET CLUSTER SETTING storage.max_sync_duration.fatal.enabled = false;
 CREATE SCHEMA IF NOT EXISTS consensus;
 CREATE SCHEMA IF NOT EXISTS adapter;
 CREATE SCHEMA IF NOT EXISTS storage;
+CREATE SCHEMA IF NOT EXISTS timestamp_oracle;
```
We should standardize on tsoracle everywhere if that's what we're going with

danhhz commented on misc/python/materialize/mzcompose/services/materialized.py:
```diff
@@ -141,6 +141,9 @@ def __init__(
                 "--persist-consensus-url=postgres://root@cockroach:26257?options=--search_path=consensus",
             ]
+            environment += [
+                "MZ_TIMESTAMP_ORACLE_URL=postgres://root@cockroach:26257?options=--search_path=timestamp_oracle"
```
ooc, why is this an env var instead of cli arg? seems like it should be the latter if possible

aljoscha replied:
It mostly is, but test infra runs older versions of materialize against newer ones to find regressions. And I think this part of test infra can't conditionally pick cli args. The old version would choke on the unknown cli arg, see https://github.com/MaterializeInc/materialize/pull/22262/commits/f457a5bfd2bdeb0bddc9c1c25a7e7abf2f4031b2 for some context.

danhhz replied:
ah, gotcha!

jkosh44 commented on src/adapter/src/coord/timeline.rs:
```diff
@@ -216,15 +215,17 @@ impl Coordinator {
                 now
             } else {
-                NOW_ZERO.clone()
+                // Timelines that are not `EpochMilliseconds` are initialized to
+                // the smallest possible timestamp.
```
I think this comment is only half accurate. The `now_fn` is also used every time we get a write timestamp after the timeline is initialized. I've struggled in the past to explain this, but basically the reason we use the minimum is because the non-`EpochMilliseconds` timelines don't have an external concept of now that all write timestamps need to be advanced to.

jkosh44 commented on src/adapter/src/coord/timestamp_selection.rs:
```diff
@@ -506,7 +506,7 @@ impl Coordinator {
             // This should only happen for literals that have no sources or sources that
             // are known to have completed (non-tailed files for example).
-            Timestamp::MAX
+            i64::MAX.try_into().expect("known to fit")
```
I think we should either add a comment or new-type to explain why `i64` is being used.

jkosh44 replied:
I'm also somewhat worried that there are other places where a timestamp can advance past `i64::MAX`. What about in the `write_ts` method itself?

aljoscha replied:
Oh yeah, this is just to try and pinpoint where the problem is. Sorry I wasn't clear about that! 🙈

danhhz reviewed (approved):
LGTM, modulo I haven't thought about the details of the u64<->i64 stuff much because I don't have a ton of context there (on board with it conceptually though), will leave the signoff on that to Joe

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=22262) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=22262) before we can accept your contribution.<br/>**12** out of **13** committers have signed the CLA.<br/><br/>:white_check_mark: philip-stoev<br/>:white_check_mark: teskje<br/>:white_check_mark: jkosh44<br/>:white_check_mark: def-<br/>:white_check_mark: moulimukherjee<br/>:white_check_mark: mjibson<br/>:white_check_mark: ParkMyCar<br/>:white_check_mark: guswynn<br/>:white_check_mark: danhhz<br/>:white_check_mark: joacoc<br/>:white_check_mark: nrainer-materialize<br/>:white_check_mark: aljoscha<br/>:x: materialize-bot<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MaterializeInc/materialize?pullRequest=22262) it.</sub>

aljoscha commented:
@danhhz & @jkosh44 I changed the oracle to use `Decimal` and I added unit tests that cover both the pg oracle and the existing oracle. @jkosh44 I had to fix up one thing in the legacy oracle to make them work, could you take a look at that? It's in the last commit.

@philip-stoev is this good to merge from qa side or do you want to, say, run nighly tests or something?

philip-stoev commented:
I have started a fresh Nightly.

jkosh44 commented on src/adapter/src/coord/timestamp_oracle/catalog_oracle.rs:
```diff
@@ -217,9 +217,12 @@ where
 
     async fn read_ts(&self) -> T {
         let ts = self.timestamp_oracle.read_ts();
+        // WIP(joe): confirm!
```
The change is from `less_than` to `less_equal`, right? Looking at the catalog oracle, this doesn't seem correct: 

```Rust
    async fn maybe_allocate_new_timestamps(&mut self, ts: &T) {
        if self.durable_timestamp.less_equal(ts)
            // Since the timestamp is at its max value, we know that no other Coord can
            // allocate a higher value.
            && self.durable_timestamp.less_than(&T::maximum())
        {
            self.durable_timestamp = ts.step_forward_by(&self.persist_interval);
            let res = self
                .timestamp_persistence
                .persist_timestamp(self.durable_timestamp.clone())
                .await;

            res.unwrap_or_terminate("can't persist timestamp");
        }
    }
```

Specifically: `if self.durable_timestamp.less_equal(ts)`.

When a new oracle starts up, I it uses the persisted timestamp exactly without incrementing it. Meaning that the persisted timestamp must always be higher than served timestamps.

Which tests required this change?

jkosh44 replied:
Ah right, that makes sense. The `Timestamp::MAX` scenario really throws a wrench into a lot of things. I think this is correct then.

def- commented:
I expected the benchmarks to work with my latest commit, but am seeing this in the logs:
```
scalability-materialized-1  | environmentd: 2023-11-21T12:37:53.973412Z  WARN environmentd::run:environmentd::serve:catalog::open:catalog::initialize_state:load_system_configuration: mz_adapter::catalog::open: cannot load unknown system parameter from stash name=timestamp_oracle
```
@aljoscha Is there something missing for parsing the system parameter?

aljoscha commented:
@def- I can't currently say why that is, but I'm seeing this in the logs of the timed-out/failed tests:
```
scalability-materialized-1  | environmentd: 2023-11-21T13:50:58.471335Z  INFO mz_adapter::coord::timeline: opening a new Postgres TimestampOracle for timeline EpochMilliseconds
```

which is not what we would want to see

aljoscha commented:
@def- it works with https://github.com/MaterializeInc/materialize/pull/22262/commits/350e35ee086246e51d3d6370c0674b3f083f462e on top, so something must not work with those parameters. Looking into it

aljoscha commented:
@def- I believe I found it: https://github.com/MaterializeInc/materialize/pull/22262/commits/20366cb3d4faccf193b557e4c0120e9ba6755e0e

def- reviewed (approved):
Sorry, missed that! Lgtm
